### PR TITLE
NumberControl: Allow empty values

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Related: https://github.com/WordPress/gutenberg/issues/33319

## Description

This fixes an issue caused by the `NumberControl` effectively enforcing a numerical value only via its use of `roundClamp`.

An example use case where an empty value would be valid is with the `BoxControl` used for padding and margin block support styles. There is a need to be able to clear a field to an undefined value.

#### Proposed solution

To bring the `NumberControl` more in line with an HTML number input this PR utilises a `required` prop to determine if an empty value can be applied via its state reducer. HTML number inputs, when not required, deem an empty value as valid.

As the `NumberControl` is currently experimental, the new `required` prop will default to false. It is currently only used directly in a few places. It does form the basis for the `UnitControl` however from what I can see the existing uses of `UnitControl` are ok allowing the field to be cleared.

## How has this been tested?

`npm run test-unit:watch packages/components/src/number-control/test`

#### Test instructions

1. Checkout this PR and run above tests
2. Run `npm run storybook:dev`
3. Open up the local storybook page for the `NumberControl`, set a placeholder and toggle off `required`. Or, follow this [link](http://localhost:50240/?path=/story/components-numbercontrol--default&knob-disabled=&knob-hideLabelFromVision=&knob-isPressEnterToChange=&knob-isShiftStepEnabled=true&knob-label=Number&knob-min=0&knob-max=100&knob-placeholder=Empty&knob-required=&knob-shiftStep=10&knob-step=1).
4. Place your cursor in the `NumberControl`, using the keyboard delete the value, then press enter. The placeholder should remain after you click elsewhere.
5. Toggle the `required` knob back on and repeat step 4. The empty value will be replaced with `0`

## Screenshots 

NumberControl

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Review by andrewserong — APPROVED** [review]
This is working nicely for me @aaronrobertshaw, tests pass, and it works according to the instructions in Storybook. I like that it's using the `required` param to enable the behaviour, too 👍

> Fixes: #33319

Tiny nit here: this PR adds the ability to fix that one, but doesn't update the usage of the UnitControl in the BoxControl to add in `required={ false }`. Did you want to update the usage here, or once this PR lands, update #33444 to use the new param?

This is working nicely for me as-is, and the logic looks straightforward, so I'll give it the ✅.

Also CC: @aristath since you've done a bit of work on this recently, too 🙂

**Comment by aaronrobertshaw** [comment]
Thanks for taking a look at this 👍 

> I like that it's using the required param to enable the behaviour, too 👍

Hopefully, this approach aligns it better with the HTML number input and makes it a little more intuitive.

> Tiny nit here: this PR adds the ability to fix that one, but doesn't update the usage of the UnitControl in the BoxControl to add in required={ false }. Did you want to update the usage here, or once this PR lands, update #33444 to use the new param?

Nice catch. I removed the updates to those components so that this PR could focus only on how the `NumberControl` allows for the empty value. I've updated the PR description to reflect this and include a "Next Steps" section.

I think we can address updating those components and any others that need it once this gets merged. How they get updated might change based our what the final decision is regarding whether the `NumberControl` should be optional or required by default.

Excluding them from #33444 also then means that PR only addresses a single issue as well.

**Comment by stokesman** [comment]
>Does it make sense to leave the NumberControl essentially being required by default, given that is its current behaviour?

It would make sense if `NumberControl` were stable. Since it’s not, I think we should change it to not required by default.

**Comment by aaronrobertshaw** [comment]
> It would make sense if NumberControl were stable. Since it’s not, I think we should change it to not required by default.

Good point. I've updated the `NumberControl` to be optional by default as well as added an extra test to cover the change. The PR description has been updated to reflect the current state of the PR.

Now the `NumberControl` is not required by default I took a look at its existing uses. These included:
- QueryToolbar
- AnglePickerControl
- RangeControl
- UnitControl

The UnitControl currently has fairly wide spread use:

- Border radius control
- Box control
- Letter spacing control
- Border width hook & global styles controls
- Layout block supports hook
- Column block width control
- Cover block min height control
- Search block width control
- Focal point picker control
- Font size picker control

All of these uses appear ok with the `NumberControl` defaulting the `required` prop to `false`.

While testing these I did find an issue with clearing the FocalPointPicker control however it turns out it was already fixed in 
https://github.com/WordPress/gutenberg/pull/33637. As such I've also rebased this PR to bring in that fix.

**Review by stokesman — APPROVED** [review]
LGTM

**Pull request merged**
